underfs/s3a/src/main/java/alluxio/underfs/s3a/S3AUnderFileSystem.java

```diff
@@ -199,32 +206,69 @@ public static S3AUnderFileSystem createInstance(AlluxioURI uri,
       clientConf.setSignerOverride(conf.get(PropertyKey.UNDERFS_S3_SIGNER_ALGORITHM));
     }
 
-    AmazonS3Client amazonS3Client = new AmazonS3Client(credentials, clientConf);
-
-    // Set a custom endpoint.
-    if (conf.isSet(PropertyKey.UNDERFS_S3_ENDPOINT)) {
-      amazonS3Client.setEndpoint(conf.get(PropertyKey.UNDERFS_S3_ENDPOINT));
+    AmazonS3ClientBuilder clientBuilder = AmazonS3ClientBuilder
+        .standard()
+        .withCredentials(credentials)
+        .withClientConfiguration(clientConf)
+        .withPathStyleAccessEnabled(Boolean.parseBoolean(conf
+            .get(PropertyKey.UNDERFS_S3_DISABLE_DNS_BUCKETS)));
+
+    AwsClientBuilder.EndpointConfiguration endpointConfiguration
+        = createEndpointConfiguration(conf, clientConf);
+    if (endpointConfiguration != null) {
+      clientBuilder.withEndpointConfiguration(endpointConfiguration);
     }
 
-    // Disable DNS style buckets, this enables path style requests.
-    if (Boolean.parseBoolean(conf.get(PropertyKey.UNDERFS_S3_DISABLE_DNS_BUCKETS))) {
-      S3ClientOptions clientOptions = S3ClientOptions.builder().setPathStyleAccess(true).build();
-      amazonS3Client.setS3ClientOptions(clientOptions);
-    }
+    AmazonS3 amazonS3Client = clientBuilder.build();
 
     ExecutorService service = ExecutorServiceFactories
         .fixedThreadPool("alluxio-s3-transfer-manager-worker",
             numTransferThreads).create();
 
     TransferManager transferManager = TransferManagerBuilder.standard()
-        .withS3Client(amazonS3Client).withExecutorFactory(() -> service)
+        .withS3Client(clientBuilder.build()).withExecutorFactory(() -> service)
         .withMultipartCopyThreshold(MULTIPART_COPY_THRESHOLD)
         .build();
 
     return new S3AUnderFileSystem(uri, amazonS3Client, bucketName,
         service, transferManager, conf, streamingUploadEnabled);
   }
 
+  /**
+   * Creates an endpoint configuration.
+   *
+   * @param conf the aluxio conf
+   * @param clientConf the aws conf
+   * @return the endpoint configuration
+   */
+  @Nullable
+  private static AwsClientBuilder.EndpointConfiguration createEndpointConfiguration(
+      UnderFileSystemConfiguration conf, ClientConfiguration clientConf) {
+    if (!conf.isSet(PropertyKey.UNDERFS_S3_ENDPOINT)) {
+      LOG.debug("No endpoint configuration generated, using default s3 endpoint");
+      return null;
+    }
+    String endpoint = conf.get(PropertyKey.UNDERFS_S3_ENDPOINT);
+    final URI epr = RuntimeHttpUtils.toUri(endpoint, clientConf);
+    LOG.debug("Creating endpoint configuration for {}", epr);
+
+    String region;
+    if (conf.isSet(PropertyKey.UNDERFS_S3_ENDPOINT_REGION)) {
+      region = conf.get(PropertyKey.UNDERFS_S3_ENDPOINT_REGION);
+    } else if (!ServiceUtils.isS3USStandardEndpoint(endpoint)) {
```

Comment:
Could you explain this a little bit more in comments?

Reply:
Yeah, reorganize the code to make it more clear.
If ends with s3.amazonaws.com, that's the standard endpoint with default region so no need to set region